File: kotlin/src/main/kotlin/advertisement/domain/model/value_object/Description.kt
package advertisement.domain.model.value_object

import advertisement.domain.errors.DescriptionEmptyError
import advertisement.domain.errors.DescriptionTooLongError
import arrow.core.Either
import arrow.core.left
import arrow.core.right
import common.domain.DomainError

class Description private constructor (private val value: String) {

    companion object {
        fun build(value: String): Either<DomainError, Description> =
            when {
                value.isEmpty() -> DescriptionEmptyError.build(value).left()
                value.length > 200 -> DescriptionTooLongError.withLongitudeMessage(value).left()
                else -> Description(value).right()
            }
    }

    fun value(): String = value
}

File: typescript/src/advertisement/domain/model/value-object/Description.ts
import { DescriptionEmptyException } from "../../exceptions/DescriptionEmptyException";
import { DescriptionTooLongException } from "../../exceptions/DescriptionTooLongException";
import { DomainException } from "../../../../common/domain/DomainException";
import * as E from '@effect-ts/core/Either';

export class Description {

    private constructor(
      readonly _value: string,
    ) {}

    public static build(value: string): E.Either<DomainException, Description> {
        const validation = this.validate(value);
        if (E.isLeft(validation))
            return E.left(validation.left);

        return E.right(new Description(value));
    }

    private static validate(value: string): E.Either<DomainException, void> {
        if (value.length === 0) {
            return E.left(DescriptionEmptyException.build());
        }

        if (value.length > 200) {
            return E.left(DescriptionTooLongException.withLongitudeMessage(value));
        }

        return E.right(undefined);
    }

    public value(): string {
        return this._value;
    }
}